### PR TITLE
Bug fix/fixes 02042020

### PR DESCRIPTION
**Pull request by jsteemann**
### Scope & Purpose

Various changes:
* added overwriteModes: "ignore" and "conflict" for AQL INSERT and non-AQL inserts.
* added exhaustive tests for overwriteModes.
* simplify all internal code related to OperationOptions handling. 
* simplify internal insert overwrite handling code.
* fix a splice-subquery related bug in explainer for subqueries that don't return any data.
* make debug dump restoration more useful.
* fix an incorrect assertion in the agency.
* display server role when using scripts/startLocalCluster.sh.

Docs PR: https://github.com/arangodb/docs/pull/405
Enterprise companion PR: https://github.com/arangodb/enterprise/pull/427

- [x] Bug-Fix for *devel-branch* (i.e. no need for backports?)
- [ ] Bug-Fix for a *released version* (did you remember to port this to all relevant release branches?)
- [ ] Strictly *new functionality* (i.e. a new feature / new option, no need for porting)
- [x] The behavior in this PR can be (and was) *manually tested* (support / qa / customers can test it)
- [ ] The behavior change can be verified via automatic tests

### Testing & Verification

This PR adds tests that were used to verify all changes:

- [ ] Added **Regression Tests** (Only for bug-fixes) 
- [ ] Added new C++ **Unit Tests** (Either GoogleTest or Catch-Test)
   - Did you add tests for a new *RestHandler* subclass ?
   - Did you add new mocks of underlying code layers to be able to verify your functionality ? 
   - ...
- [ ] Added new **integration tests** (i.e. in shell_server / shell_server_aql)
- [ ] Added new **resilience tests** (only if the feature is impacted by failovers)

Additionally:

http://172.16.10.101:8080/view/PR/job/arangodb-matrix-pr/9287/

### CODE REVIEW

**Comment by jsteemann** [comment]
http://172.16.10.101:8080/view/PR/job/arangodb-matrix-pr/9289/

**Comment by jsteemann** [comment]
http://172.16.10.101:8080/view/PR/job/arangodb-matrix-pr/9309/

**Comment by jsteemann** [comment]
http://172.16.10.101:8080/view/PR/job/arangodb-matrix-pr/9311/

File: CHANGELOG [hunk 1]
```diff
@@ -8,6 +8,46 @@ devel
...
+  by specifying it in the INSERT's `OPTIONS`, e.g.
+
+      INSERT { _key: ..., .... } INTO collection OPTIONS { overwriteMode: "ignore" }
+
+  Again, when the overwrite mode "ignore" is used from AQL, it does not
+  support returning the old or new document versions. Using "RETURN OLD" in 
+  an INSERT operation that uses the "ignore" overwrite mode will trigger
+  a parse error, as there will be no old version returned, and "RETURN NEW"
+  will only return the document in case it was inserted. In case the document
+  already existed, "RETURN NEW" will return "null".
```
Comment (Simran-B) [inline CHANGELOG:43]: If AQL either returns the new document or null in the "ignore" case, why does the API not do the same for `returnNew`?

Comment (jsteemann) [reply CHANGELOG:43]: @Simran-B : Doesn't it? Can you elaborate what the difference is?

**Pull request merged**